// bottleneck/rollup.config.light.js
// Import required plugins for Rollup bundler
import commonjs from 'rollup-plugin-commonjs';
import json from 'rollup-plugin-json';
import resolve from 'rollup-plugin-node-resolve';

// Define the banner text for the top of the output file
const bannerLines = [
  'This file contains the Bottleneck library (MIT), compiled to ES2017, and without Clustering support.',
  'https://github.com/SGrondin/bottleneck',
].map(x => `  * ${x}`).join('\n');
const banner = `/**\n${bannerLines}\n  */`;

// Define the missing export message and the list of files to exclude
const missing = `export default () => console.log('You must import the full version of Bottleneck in order to use this feature.');`;
const exclude = [
  'RedisDatastore.js',
  'RedisConnection.js',
  'IORedisConnection.js',
  'Scripts.js'
];

// Configuration for Rollup bundler
export default {
  // Input file for the bundler
  input: 'lib/index.js',
  // Output file for the bundler
  output: {
    // Name of the exported object
    name: 'Bottleneck',
    // Output file path
    file: 'light.js',
    // Disable source maps
    sourcemap: false,
    // Globals to be defined in the output file
    globals: {},
    // Output format
    format: 'umd',
    // Banner text for the top of the output file
    banner
  },
  // List of external dependencies
  external: [],
  // Plugins for the bundler
  plugins: [
    // Plugin to handle JSON modules
    json(),
    // Plugin to load external modules
    {
      load: id => {
        // Get the file name from the module ID
        const chunks = id.split('/');
        const file = chunks[chunks.length - 1];
        // Check if
