// @octokit/oauth-methods/dist-src/delete-token.js
import { request as defaultRequest } from "@octokit/request";
import btoa from "btoa-lite";

// deleteToken function deletes a token for a specified client ID
async function deleteToken(options) {
  // Ensure a custom request is provided in tests
  const request = options.request || /* istanbul ignore next: we always pass a custom request in tests */
  defaultRequest;

  // Combine the client ID and client secret using btoa for base64 encoding
  const auth = btoa(`${options.clientId}:${options.clientSecret}`);

  // Make a DELETE request to the GitHub API to revoke the token
  return request(
    "DELETE /applications/{client_id}/token",
    {
      // Include the authorization header with the base64 encoded client ID and client secret
      headers: {
        authorization: `basic ${auth}`
      },
      // Include the client ID in the path
      client_id: options.clientId,
      // Include the token to be revoked as a parameter
      access_token: options.token
    }
  );
}

// Export the deleteToken function for use in other modules
export {
  deleteToken
};
