// @octokit/oauth-app/dist-src/methods/create-token.js
import * as OAuthAppAuth from "@octokit/auth-oauth-app";
import { emitEvent } from "../emit-event.js";

// createTokenWithState is an asynchronous function that creates a new OAuth token with the given state and options.
async function createTokenWithState(state, options) {
  // It first authenticates the user using the provided state and options.
  const authentication = await state.octokit.auth({
    type: "oauth-user",
    ...options // Spread operator is used to pass the options object as arguments to the auth method.
  });

  // After successful authentication, it emits a 'token' event with the following details:
  // - The created token
  // - The scopes of the token
  // - The authentication object containing token and scopes
  // - A new Octokit instance with the created token and scopes
  await emitEvent(state, {
    name: "token",
    action: "created",
    token: authentication.token,
    scopes: authentication.scopes,
    authentication,
    octokit: new state.Octokit({
      authStrategy: OAuthAppAuth.createOAuthUserAuth,
      auth: {
        clientType: state.clientType,
        clientId: state.clientId,
        clientSecret: state.clientSecret,
        token: authentication.token,
        scopes: authentication.scopes,
        refreshToken: authentication.refreshToken,
        expiresAt: authentication.expiresAt,
        refreshTokenExpiresAt: authentication.refreshTokenExpiresAt
      }
    })
  });

  // Finally, it returns an object containing the authentication object.
  return { authentication };
}

// Exports the createTokenWithState function for use in other modules.
export { createTokenWithState };
